contracts/zero-ex/contracts/src/features/MetaTransactions.sol

```diff
@@ -0,0 +1,412 @@
+        state.hash = getMetaTransactionHash(mtx);
+        state.mtx = mtx;
+        state.signature = signature;
+
+        _validateMetaTransaction(state);
+
+        // Mark the transaction executed.
+        assert(block.number > 0);
+        LibMetaTransactionsStorage.getStorage()
+            .mtxHashToExecutedBlockNumber[state.hash] = block.number;
+
+        // Execute the call based on the selector.
+        state.selector = mtx.callData.readBytes4(0);
+        if (state.selector == ITransformERC20.transformERC20.selector) {
+            returnData = _executeTransformERC20Call(state);
+        } else {
+            LibMetaTransactionsRichErrors
+                .MetaTransactionUnsupportedFunctionError(state.hash, state.selector)
+                .rrevert();
+        }
+        // Pay the fee to the sender.
+        if (mtx.feeAmount > 0) {
+            ITokenSpender(address(this))._spendERC20Tokens(
+                mtx.feeToken,
+                mtx.signer, // From the signer.
+                sender, // To the sender.
+                mtx.feeAmount
+            );
+        }
+        emit MetaTransactionExecuted(
+            state.hash,
+            state.selector,
+            mtx.signer,
+            mtx.sender
+        );
+    }
+
+    /// @dev Validate that a meta-transaction is executable.
+    function _validateMetaTransaction(ExecuteState memory state)
+        private
+        view
+    {
+        // Must be from the required sender, if set.
+        if (state.mtx.sender != address(0) && state.mtx.sender != state.sender) {
+            LibMetaTransactionsRichErrors
+                .MetaTransactionWrongSenderError(
+                    state.hash,
+                    state.sender,
+                    state.mtx.sender
+                ).rrevert();
+        }
+        // Must not be expired.
+        if (state.mtx.expirationTime <= block.timestamp) {
+            LibMetaTransactionsRichErrors
+                .MetaTransactionExpiredError(
+                    state.hash,
+                    block.timestamp,
+                    state.mtx.expirationTime
+                ).rrevert();
+        }
+        // Must have a valid gas price.
+        if (state.mtx.minGasPrice > tx.gasprice || state.mtx.maxGasPrice < tx.gasprice) {
+            LibMetaTransactionsRichErrors
+                .MetaTransactionGasPriceError(
+                    state.hash,
+                    tx.gasprice,
+                    state.mtx.minGasPrice,
+                    state.mtx.maxGasPrice
+                ).rrevert();
+        }
+        // Must have enough ETH.
+        state.selfBalance  = address(this).balance;
+        if (state.mtx.value > state.selfBalance) {
+            LibMetaTransactionsRichErrors
+                .MetaTransactionInsufficientEthError(
+                    state.hash,
+                    state.selfBalance,
+                    state.mtx.value
+                ).rrevert();
+        }
+        // Must be signed by signer.
```

Comment:
should we skip the signature check if `state.sender == state.mtx.signer`?

Reply:
Good point. We do skip it in the Exchange. I wonder if that's a vulnerability though since you can trick a mtx relayer into executing a transaction as itself, if they aren't paying attention.

Reply:
yeah that might be true –– in that case I think we should just add a generic `batchExecute` feature